### PR TITLE
Field effect system

### PR DESCRIPTION
**Pull request by Aeliko**
- Implement Grassy Terrain, Electric Terrain, Misty Terrain, Psychic Terrain
- Implement way to add custom effects before move is queued (`345.s`)
- Implement way to add custom effects at the first turn (`346.s`)
- Implement way to add custom effects at the end of the turn (`347.s`)
- Implement "framework" to add "field effects"
- Implement toggle to turn off "field effects" (default off)

### CODE REVIEW

File: armips/move/battle_eff_seq/155.s [hunk 1]
```diff
@@ -6,12 +6,15 @@
...
 .include "armips/include/movenums.s"
+.include "armips/include/constants.s"
 
 .create "build/move/battle_eff_seq/0_155", 0
 
 a030_155:
     ifmonstat IF_MASK, BATTLER_ATTACKER, MON_DATA_STATUS_2, 0x1000, _00B4
     checkitemeffect 0x0, BATTLER_ATTACKER, 0x63, _0064
+    // skip charge turn while in cave
+    ifcurrentfieldistype TERRAIN_CAVE, _0064
```
Comment (BluRosie) [inline armips/move/battle_eff_seq/155.s:17]: please remove custom effects (if you want i can do it)

File: armips/move/battle_eff_seq/263.s [hunk 2]
```diff
@@ -6,12 +6,15 @@
...
 .include "armips/include/movenums.s"
+.include "armips/include/constants.s"
 
 .create "build/move/battle_eff_seq/0_263", 0
 
 a030_263:
     ifmonstat IF_MASK, BATTLER_ATTACKER, MON_DATA_STATUS_2, 0x1000, _00B4
     checkitemeffect 0x0, BATTLER_ATTACKER, 0x63, _0064
+    // skip charge turn when in cave
+    ifcurrentfieldistype TERRAIN_CAVE, _0064
```
Comment (BluRosie) [inline armips/move/battle_eff_seq/263.s:17]: custom effects

File: armips/move/battle_sub_seq/217.s [hunk 3]
```diff
@@ -20,10 +21,12 @@ _0030:
     wait 0x1E
     setstatus2effect BATTLER_ATTACKER, 0xA
     waitmessage
+    checkitemeffect 0x1, BATTLER_ATTACKER, 0x63, SkipRemoveItem
```
Comment (BluRosie) [inline armips/move/battle_sub_seq/217.s:24]: more custom effect stuff

File: armips/move/battle_sub_seq/316.s [hunk 4]
```diff
@@ -36,6 +38,23 @@ StatusEffect:
...
+    gotosubscript 12
+    //def up
+    changevar VAR_OP_SET, VAR_ADD_EFFECT_ATTRIBUTE, DEFENCE_UP_2
+    gotosubscript 12
+    //acc up
+    changevar VAR_OP_SET, VAR_ADD_EFFECT_ATTRIBUTE, ACCURACY_UP_2
+    gotosubscript 12
+    changevar VAR_OP_CLEARMASK, VAR_SERVER_STATUS2, 0x2
+    changevar VAR_OP_CLEARMASK, VAR_SERVER_STATUS2, 0x80
+    endscript
```
Comment (BluRosie) [inline armips/move/battle_sub_seq/316.s:57]: does coil raise stats 2x in grassy terrain?  i don't actually know lol

File: armips/move/battle_sub_seq/331.s [hunk 5]
```diff
@@ -6,11 +6,13 @@
 .include "armips/include/itemnums.s"
 .include "armips/include/monnums.s"
 .include "armips/include/movenums.s"
+.include "armips/include/constants.s"
 
 .create "build/move/battle_sub_seq/1_331", 0
 
 growthSubEffectScript: // a001_331
     if IF_MASK, VAR_FIELD_EFFECT, 0x30, SunIsUp // checks for sun weather
+    ifcurrentfieldistype TERRAIN_GRASS, SunIsUp // gain double of the normal stat-changing effects in grass terrain
```
Comment (BluRosie) [inline armips/move/battle_sub_seq/331.s:15]: same here as coil

File: armips/move/battle_sub_seq/345.s [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+            endscript
+        GrassFieldSoftenNerf:
+            changevar VAR_OP_SET, VAR_DAMAGE_MULT, 0x5 // 0.5 multiplier
+            printattackmessage
+            waitmessage
+            wait 0x1E
+            printmessage GrassFieldSoftenNerfMsg, TAG_NONE, "NaN", "NaN", "NaN", "NaN", "NaN", "NaN" // The grass softened the attack...
+            waitmessage
+            wait 0x1E
+            endscript
```
Comment (BluRosie) [inline armips/move/battle_sub_seq/345.s:124]: are these all vanilla interactions?

Comment (BluRosie) [reply armips/move/battle_sub_seq/345.s:124]: same with all the cool interactions
while they are good ideas, i don't mean to merge them into an engine that is aiming to implement parity with later generations

File: armips/move/battle_sub_seq/346.s [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+IsGrassField:
+    printmessage GrassFieldInitialMsg, TAG_NONE, "NaN", "NaN", "NaN", "NaN", "NaN", "NaN" // The field is in full bloom.
+    waitmessage
+    wait 0x1E
+    endscript
+IsCaveField:
+    printmessage CaveFieldInitialMsg, TAG_NONE, "NaN", "NaN", "NaN", "NaN", "NaN", "NaN" // The cave echoes dully...
+    waitmessage
+    wait 0x1E
+    endscript
```
Comment (BluRosie) [inline armips/move/battle_sub_seq/346.s:31]: just jotting more things

File: data/text/197.txt [hunk 8]
```diff
@@ -1386,3 +1386,128 @@ The foe’s {STRVAR_1 1, 0, 0}’s {STRVAR_1 5, 1, 0}\nraised its Speed!
...
+The field is in full bloom.
+The grass below caught flame!
+The wind picked up strength\nfrom the field!
+The cave echoes dully...
+ECHO-Echo-echo!
+The cave choked out the air!
+The cavern strengthened\nthe attack!
+...Piled on!
+Bits of rock fell from the crumbling ceiling!
+The quake collapsed the ceiling!
```
Comment (BluRosie) [inline data/text/197.txt:1513]: All this stuff too

File: include/battle.h [hunk 9]
```diff
@@ -1207,6 +1213,8 @@ struct __attribute__((packed)) newBattleStruct
     CATS_ACT_PTR WeatherOAM;
     SysTask *weatherUpdateTask;
     u32 weather;
+
+    TerrainOverlay terrainOverlay;
```
Comment (BluRosie) [inline include/battle.h:1217]: could potentially just put this in the main BattleStruct

File: include/debug.h [hunk 10]
```diff
@@ -45,6 +45,6 @@
 //#define DEBUG_CAPTURE_RATE_PERCENTAGES
 
 // DEBUG_SERVER_QUEUE prints debug information about the status of the server queue when the waitmessage script command
-//#define DEBUG_SERVER_QUEUE
+// #define DEBUG_TERRAIN
```
Comment (BluRosie) [inline include/debug.h:48]: need your own debug define for terrain (if you use this at all)

File: src/battle/ability.c [hunk 11]
```diff
@@ -111,6 +111,93 @@ const u16 PowderMoveList[] = {
...
+         || (sp->battlemon[client_no].effect_of_moves & MOVE_EFFECT_FLAG_INGRAIN)  // is Ingrained
+         || (sp->field_condition & FIELD_STATUS_GRAVITY))) {
+        // not in a semi-vulnerable state
+        if ((sp->battlemon[client_no].effect_of_moves & (MOVE_EFFECT_FLAG_FLYING_IN_AIR | MOVE_EFFECT_FLAG_DIGGING | MOVE_EFFECT_FLAG_IS_DIVING | MOVE_EFFECT_FLAG_SHADOW_FORCE)) == 0) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
```
Comment (BluRosie) [inline src/battle/ability.c:200]: potentially just do like.  IsClientGrounded or something, it's a general-use func that can be used wherever

**Review by BluRosie — CHANGES_REQUESTED** [review]
how do you manage damage calc modifications for each terrain?  also jotted down a bunch that isn't mainline behavior so it isn't suitable for hg-engine outside of a personal hack.  i can deal with the things i notice if you don't feel like doing it, i guess just let me know haha

**Comment by BluRosie** [comment]
hoping to review this tonight and get it merged!!

**Review by NathanWichers — APPROVED** [review]
looks good to me!  anything else you want to overhaul?

**Review by BluRosie — APPROVED** [review]

**Pull request merged**